params/precompile_config.go

```diff
@@ -0,0 +1,273 @@
+)
+
+// precompileKey is a helper type used to reference each of the
+// possible stateful precompile types that can be activated
+// as a network upgrade.
+type precompileKey int
+
+const (
+	contractDeployerAllowListKey precompileKey = iota + 1
+	contractNativeMinterKey
+	txAllowListKey
+	feeManagerKey
+)
+
+var (
+	precompileKeys = []precompileKey{contractDeployerAllowListKey, contractNativeMinterKey, txAllowListKey, feeManagerKey}
+)
+
+// PrecompileUpgrade is a helper struct embedded in UpgradeConfig, representing
+// each of the possible stateful precompile types that can be activated
+// as a network upgrade.
+type PrecompileUpgrade struct {
+	ContractDeployerAllowListConfig *precompile.ContractDeployerAllowListConfig `json:"contractDeployerAllowListConfig,omitempty"` // Config for the contract deployer allow list precompile
+	ContractNativeMinterConfig      *precompile.ContractNativeMinterConfig      `json:"contractNativeMinterConfig,omitempty"`      // Config for the native minter precompile
+	TxAllowListConfig               *precompile.TxAllowListConfig               `json:"txAllowListConfig,omitempty"`               // Config for the tx allow list precompile
+	FeeManagerConfig                *precompile.FeeConfigManagerConfig          `json:"feeManagerConfig,omitempty"`                // Config for the fee manager precompile
+}
+
+func (p *PrecompileUpgrade) getByKey(key precompileKey) (precompile.StatefulPrecompileConfig, bool) {
+	switch key {
+	case contractDeployerAllowListKey:
+		return p.ContractDeployerAllowListConfig, p.ContractDeployerAllowListConfig != nil
+	case contractNativeMinterKey:
+		return p.ContractNativeMinterConfig, p.ContractNativeMinterConfig != nil
+	case txAllowListKey:
+		return p.TxAllowListConfig, p.TxAllowListConfig != nil
+	case feeManagerKey:
+		return p.FeeManagerConfig, p.FeeManagerConfig != nil
+	default:
+		panic(fmt.Sprintf("unknown upgrade key: %v", key))
+	}
+}
+
+// VerifyPrecompileUpgrades checks [c.PrecompileUpgrades] is well formed:
+// - [upgrades] must specify exactly one key per PrecompileUpgrade
+// - the specified blockTimestamps must monotonically increase
+// - the specified blockTimestamps must be compatible with those
+//   specified in the chainConfig by genesis.
+// - check a precompile is disabled before it is re-enabled
+func (c *ChainConfig) VerifyPrecompileUpgrades() error {
+	for i, upgrade := range c.PrecompileUpgrades {
+		hasKey := false // used to verify if there is only one key per Upgrade
+
+		for _, key := range precompileKeys {
+			config, ok := upgrade.getByKey(key)
+			if !ok {
+				continue
+			}
+			if hasKey {
+				return fmt.Errorf("PrecompileUpgrades[%d] has more than one key set", i)
+			}
+			if config.Timestamp() == nil {
+				return fmt.Errorf("PrecompileUpgrades[%d] cannot have a nil timestamp", i)
+			}
+			hasKey = true
+		}
+		if !hasKey {
+			return fmt.Errorf("empty precompile upgrade at index %d", i)
+		}
+	}
+
+	for _, key := range precompileKeys {
```

Comment:
Currently we only assert configs are monotonically increasing for a single precompile key, but we should enforce that they are monotonically increasing for any precompile key as well.

Reply:
https://github.com/ava-labs/subnet-evm/pull/168 adds this verification

Reply:
Merged, thanks for putting this up!